src/power.jl

```diff
@@ -7,156 +7,251 @@
 #
 
 
-## Int power ##
-doc"""
-```
-^(a, x)
-```
-
-Return the Taylor expansion of $a^x$ for `a::Taylor1` polynomial and `x::Number`.
-If `x` is non integer and the 0-th order coefficient is zero, an
-`ArgumentError` is thrown.
-"""
-function ^{T<:Number}(a::Taylor1{T}, n::Integer)
-    n == 0 && return one(a)
-    n == 1 && return a
-    n == 2 && return square(a)
-    n < 0 && return inv( a^(-n) )
-    return power_by_squaring(a, n)
-end
-
-function ^{T<:Integer}(a::Taylor1{T}, n::Integer)
+function ^(a::HomogeneousPolynomial, n::Integer)
     n == 0 && return one(a)
     n == 1 && return a
     n == 2 && return square(a)
     n < 0 && throw(DomainError())
-    return power_by_squaring(a, n)
+    return Base.power_by_squaring(a, n)
 end
 
-## power_by_squaring; slightly modified from base/intfuncs.jl
-## Licensed under MIT "Expat"
-function power_by_squaring(x::Taylor1, p::Integer)
-    p == 1 && return copy(x)
-    p == 0 && return one(x)
-    p == 2 && return square(x)
-    t = trailing_zeros(p) + 1
-    p >>= t
-
-    while (t -= 1) > 0
-        x = square(x)
+for T in (:Taylor1, :TaylorN)
+    @eval function ^{T<:Number}(a::$T{T}, n::Integer)
+        n == 0 && return one(a)
+        n == 1 && return a
+        n == 2 && return square(a)
+        n < 0 && return inv( a^(-n) )
+        return Base.power_by_squaring(a, n)
```

Comment:
The advantage of `power_by_squaring` is that, if you have `a::Taylor1{Int}`, the result is of the same type, as it would be expected. This is not preserved by `exp(n*log(a))`, which would return `Taylor1{Float64}`.

Long time ago I did some comparisons with what we have now and the same type of implementation for `^(a,r::Real)`; `power_by_squaring was better, at least for small `n`. 

The same subtlety about the type of the result appears here. The nice thing of preserving the type is that you get exact answers for Fateman-like tests.